src/query/functions/src/aggregates/aggregate_sum_zero.rs

```diff
@@ -0,0 +1,267 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Layout;
+use std::fmt;
+use std::sync::Arc;
+
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_expression::types::number::NumberColumnBuilder;
+use databend_common_expression::types::AccessType;
+use databend_common_expression::types::ArgType;
+use databend_common_expression::types::Bitmap;
+use databend_common_expression::types::DataType;
+use databend_common_expression::types::NullableType;
+use databend_common_expression::types::NumberDataType;
+use databend_common_expression::types::UInt64Type;
+use databend_common_expression::types::UnaryType;
+use databend_common_expression::types::ValueType;
+use databend_common_expression::AggrStateRegistry;
+use databend_common_expression::AggrStateType;
+use databend_common_expression::BlockEntry;
+use databend_common_expression::ColumnBuilder;
+use databend_common_expression::ProjectedBlock;
+use databend_common_expression::Scalar;
+use databend_common_expression::StateSerdeItem;
+
+use super::assert_params;
+use super::AggrState;
+use super::AggrStateLoc;
+use super::AggregateFunction;
+use super::AggregateFunctionDescription;
+use super::AggregateFunctionSortDesc;
+use super::StateAddr;
+use crate::aggregates::assert_unary_arguments;
+
+struct AggregateSumZeroState {
+    sum: u64,
+}
+
+#[derive(Clone)]
+pub struct AggregateSumZeroFunction {
+    display_name: String,
+}
+
+impl AggregateSumZeroFunction {
+    pub fn try_create(
+        display_name: &str,
+        params: Vec<Scalar>,
+        arguments: Vec<DataType>,
+        _sort_descs: Vec<AggregateFunctionSortDesc>,
+    ) -> Result<Arc<dyn AggregateFunction>> {
+        assert_params(display_name, params.len(), 0)?;
+        assert_unary_arguments(display_name, arguments.len())?;
+        let data_type = arguments[0].clone();
+
+        if data_type.remove_nullable() != DataType::Number(NumberDataType::UInt64) {
+            return Err(ErrorCode::InvalidArgument(format!(
+                "Invalid argument type for {}, must be uint64",
+                display_name
+            )));
+        }
+
+        Ok(Arc::new(AggregateSumZeroFunction {
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> AggregateFunctionDescription {
+        let features = super::AggregateFunctionFeatures {
+            returns_default_when_only_null: true,
+            is_decomposable: true,
+            ..Default::default()
+        };
+        AggregateFunctionDescription::creator_with_features(Box::new(Self::try_create), features)
+    }
+}
+
+impl AggregateFunction for AggregateSumZeroFunction {
```

Comment:
Looks so much like sum it's hard to notice the difference, should be able to reuse the common parts

Reply:
It's used internally.

`count(x)` returns i64
Yet `sum(x)` returns nullable.

We must have a sum function that returns non-nullable results, otherwise it will get schema mismatch.